server/fleet/apple_mdm.go

```diff
@@ -188,43 +187,55 @@ type MDMAppleHostDetails struct {
 	Model        string
 }
 
-// MDMHostIngester represents the interface used to ingest an MDM device as a Fleet host pending enrollment.
-type MDMHostIngester interface {
-	Ingest(context.Context, *http.Request) error
-}
-
-// MDMAppleHostIngester implements the MDMHostIngester interface in connection with Apple MDM services.
-type MDMAppleHostIngester struct {
-	ds     Datastore
-	logger kitlog.Logger
-}
-
-// NewMDMAppleHostIngester returns a new instance of an MDMAppleHostIngester.
-func NewMDMAppleHostIngester(ds Datastore, logger kitlog.Logger) *MDMAppleHostIngester {
-	return &MDMAppleHostIngester{ds: ds, logger: logger}
-}
-
-// Ingest handles incoming http requests that follow Apple's MDM checkin protocol. For valid
-// checkin requests, Ingest decodes the XML body and ingests new host details into the associated
-// datastore. See also https://developer.apple.com/documentation/devicemanagement/check-in.
-func (ingester *MDMAppleHostIngester) Ingest(ctx context.Context, r *http.Request) error {
+// HandleMDMCheckinRequest handles incoming http requests
+// that follow Apple's MDM checkin protocol.
+//
+// - For valid Authenticate checkin requests, it decodes
+// the XML body and ingests new host details into the
+// associated datastore.
+// - For TokenUpdate and CheckOut requests it creates
+// activity logs.
+//
+// See also
+// https://developer.apple.com/documentation/devicemanagement/check-in.
+func HandleMDMCheckinRequest(ctx context.Context, r *http.Request, ds Datastore) error {
 	if isMDMAppleCheckinReq(r) {
-		host := MDMAppleHostDetails{}
-
-		ok, err := decodeMDMAppleCheckinReq(r, &host)
-		switch {
-		case err != nil:
+		msg, err := decodeMDMAppleCheckinReq(r)
+		if err != nil {
 			return fmt.Errorf("decode checkin request: %w", err)
-		case !ok:
-			return nil
-		default:
-			// continue
 		}
 
-		if err := ingester.ds.IngestMDMAppleDeviceFromCheckin(ctx, host); err != nil {
-			return err
+		switch m := msg.(type) {
+		case *mdm.Authenticate:
+			host := MDMAppleHostDetails{}
+			host.SerialNumber = m.SerialNumber
+			host.UDID = m.UDID
+			host.Model = m.Model
+			return ds.IngestMDMAppleDeviceFromCheckin(ctx, host)
+		case *mdm.TokenUpdate:
+			info, err := ds.GetHostMDMCheckinInfo(ctx, m.Enrollment.UDID)
+			if err != nil {
+				return err
+			}
+			return ds.NewActivity(ctx, nil, &ActivityTypeMDMEnrolled{
+				HostSerial:       info.HardwareSerial,
+				InstalledFromDEP: info.InstalledFromDEP,
+			})
+		case *mdm.CheckOut:
```

Comment:
Should there be any updates to the `host_mdm` table in this case?

Reply:
good point, I added logic to set `enroll = 0` on unenrollment